From: txtyb <txtybbill@163.com>
Subject: [PATCH] arm64: dts: oplus: AD296: restore ADFR tables and add SA floor table

Add qsync AVR step fps to every timing, restore the ADFR min-fps mapping
tables and command sequences for the 144Hz and 165Hz timings, and add the
oplus,adfr-sa-min-fps-bl-table used by the driver to pick the SA floor.

Port of the Project Infinity-X implementation.
---

--- a/qcom/display/oplus/panel/dsi-panel-AD296-P-3-A0020-dsc-cmd.dtsi
+++ b/qcom/display/oplus/panel/dsi-panel-AD296-P-3-A0020-dsc-cmd.dtsi
@@ -54,6 +54,7 @@
 				qcom,mdss-dsc-version = <0x12>;
 				qcom,mdss-dsi-timing-default;
 				qcom,mdss-dsi-panel-framerate = <120>;
+				qcom,dsi-qsync-mode-avr-step-fps = <120>;
 				qcom,mdss-dsi-panel-clockrate = <1112600000>;
 				qcom,mdss-dsi-panel-width = <1272>;
 				qcom,mdss-dsi-panel-height = <2772>;
@@ -656,6 +657,7 @@
 				qcom,mdss-dsc-version = <0x12>;
 				qcom,mdss-dsi-timing-default;
 				qcom,mdss-dsi-panel-framerate = <90>;
+				qcom,dsi-qsync-mode-avr-step-fps = <90>;
 				qcom,mdss-dsi-panel-clockrate = <1112600000>;
 				qcom,mdss-dsi-panel-width = <1272>;
 				qcom,mdss-dsi-panel-height = <2772>;
@@ -1253,6 +1255,7 @@
 				qcom,mdss-dsc-version = <0x12>;
 				qcom,mdss-dsi-timing-default;
 				qcom,mdss-dsi-panel-framerate = <60>;
+				qcom,dsi-qsync-mode-avr-step-fps = <60>;
 				qcom,mdss-dsi-panel-clockrate = <1112600000>;
 				qcom,mdss-dsi-panel-width = <1272>;
 				qcom,mdss-dsi-panel-height = <2772>;
@@ -1858,6 +1861,7 @@
 				qcom,mdss-dsc-version = <0x12>;
 				qcom,mdss-dsi-timing-default;
 				qcom,mdss-dsi-panel-framerate = <144>;
+				qcom,dsi-qsync-mode-avr-step-fps = <144>;
 				qcom,mdss-dsi-panel-clockrate = <1363200000>;
 				qcom,mdss-dsi-panel-width = <1272>;
 				qcom,mdss-dsi-panel-height = <2772>;
@@ -2289,6 +2293,7 @@
 				qcom,mdss-dsc-version = <0x12>;
 				qcom,mdss-dsi-timing-default;
 				qcom,mdss-dsi-panel-framerate = <165>;
+				qcom,dsi-qsync-mode-avr-step-fps = <165>;
 				qcom,mdss-dsi-panel-clockrate = <1363200000>;
 				qcom,mdss-dsi-panel-width = <1272>;
 				qcom,mdss-dsi-panel-height = <2772>;
@@ -2710,6 +2715,7 @@
 				qcom,mdss-dsc-version = <0x12>;
 				qcom,mdss-dsi-timing-default;
 				qcom,mdss-dsi-panel-framerate = <120>;
+				qcom,dsi-qsync-mode-avr-step-fps = <120>;
 				qcom,mdss-dsi-panel-clockrate = <1112600000>;
 				qcom,mdss-dsi-panel-width = <1272>;
 				qcom,mdss-dsi-panel-height = <2772>;
@@ -3154,7 +3160,9 @@
 	oplus,adfr-test-te-gpio = <&tlmm 85 0>;
 	/* just for qsync enable*/
 	qcom,qsync-enable;
-	qcom,mdss-dsi-qsync-min-refresh-rate = <30>;
+	qcom,mdss-dsi-qsync-min-refresh-rate = <1>;
+	/* backlight threshold and sa min fps pairs, highest threshold first, 0 keeps the mode rate */
+	oplus,adfr-sa-min-fps-bl-table = <1160 1>, <919 30>, <0 0>;
 
 	// qcom,dsi-dyn-clk-enable;
 
@@ -3598,11 +3606,219 @@
 			/* standard adfr */
 			qcom,disable-rsc-solver;
 			qcom,mdss-dsi-h-sync-skew = <0>;
+			oplus,adfr-min-fps-mapping-table = <144 72 48 36 24 12>;
+
+			qcom,mdss-dsi-adfr-min-fps-0-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 00 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-adfr-min-fps-1-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 01 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-adfr-min-fps-2-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 02 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-adfr-min-fps-3-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 03 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-adfr-min-fps-4-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 05 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-adfr-min-fps-5-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 0B 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+
+			qcom,mdss-dsi-hpwm-adfr-min-fps-0-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 00 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-hpwm-adfr-min-fps-1-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 01 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-hpwm-adfr-min-fps-2-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 02 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-hpwm-adfr-min-fps-3-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 03 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-hpwm-adfr-min-fps-4-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 05 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-hpwm-adfr-min-fps-5-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 0B 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+
+			qcom,mdss-dsi-bigdc-adfr-min-fps-0-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 00 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-bigdc-adfr-min-fps-1-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 01 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-bigdc-adfr-min-fps-2-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 02 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-bigdc-adfr-min-fps-3-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 03 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-bigdc-adfr-min-fps-4-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 05 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-bigdc-adfr-min-fps-5-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 87 0B 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+
+			qcom,mdss-dsi-adfr-min-fps-0-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-adfr-min-fps-1-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-adfr-min-fps-2-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-adfr-min-fps-3-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-adfr-min-fps-4-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-adfr-min-fps-5-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-hpwm-adfr-min-fps-0-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-hpwm-adfr-min-fps-1-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-hpwm-adfr-min-fps-2-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-hpwm-adfr-min-fps-3-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-hpwm-adfr-min-fps-4-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-hpwm-adfr-min-fps-5-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-bigdc-adfr-min-fps-0-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-bigdc-adfr-min-fps-1-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-bigdc-adfr-min-fps-2-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-bigdc-adfr-min-fps-3-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-bigdc-adfr-min-fps-4-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-bigdc-adfr-min-fps-5-command-state = "dsi_hs_mode";
 		};
 		timing@sdc_fhd_165 {
 			/* standard adfr */
 			qcom,disable-rsc-solver;
 			qcom,mdss-dsi-h-sync-skew = <0>;
+			oplus,adfr-min-fps-mapping-table = <165 55 33 15 11>;
+
+			qcom,mdss-dsi-adfr-min-fps-0-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 00 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-adfr-min-fps-1-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 02 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-adfr-min-fps-2-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 04 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-adfr-min-fps-3-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 0A 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-adfr-min-fps-4-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 0E 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+
+			qcom,mdss-dsi-hpwm-adfr-min-fps-0-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 00 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-hpwm-adfr-min-fps-1-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 02 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-hpwm-adfr-min-fps-2-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 04 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-hpwm-adfr-min-fps-3-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 0A 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-hpwm-adfr-min-fps-4-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 0E 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+
+			qcom,mdss-dsi-bigdc-adfr-min-fps-0-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 00 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-bigdc-adfr-min-fps-1-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 02 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-bigdc-adfr-min-fps-2-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 04 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-bigdc-adfr-min-fps-3-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 0A 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+			qcom,mdss-dsi-bigdc-adfr-min-fps-4-command = [
+				39 00 00 40 00 00 04 FF 5A A5 2D
+				39 00 00 40 00 00 03 89 0E 00
+				39 00 00 00 00 00 04 FF 5A A5 00
+			];
+
+			qcom,mdss-dsi-adfr-min-fps-0-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-adfr-min-fps-1-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-adfr-min-fps-2-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-adfr-min-fps-3-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-adfr-min-fps-4-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-hpwm-adfr-min-fps-0-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-hpwm-adfr-min-fps-1-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-hpwm-adfr-min-fps-2-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-hpwm-adfr-min-fps-3-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-hpwm-adfr-min-fps-4-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-bigdc-adfr-min-fps-0-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-bigdc-adfr-min-fps-1-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-bigdc-adfr-min-fps-2-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-bigdc-adfr-min-fps-3-command-state = "dsi_hs_mode";
+			qcom,mdss-dsi-bigdc-adfr-min-fps-4-command-state = "dsi_hs_mode";
 		};
 		timing@oplus_fhd_120 {
 			/* oplus adfr */
